### PR TITLE
access log: added new START_TIME_LOCAL & EMIT_TIME_LOCAL command

### LINKED ISSUE #35116: proposal: introduce START_TIME_LOCAL and EMIT_TIME_LOCAL for local time zone in access_log
_Originally posted in https://github.com/envoyproxy/envoy/issues/23260#issuecomment-2122534085_

The current time format is already complicated, i.e.
```
%START_TIME(%Y/%m/%dT%H:%M:%S%z)%
%START_TIME(%s.%3f)%
```

New command operators might be simpler, similar to `time_local` in [NGINX](https://nginx.org/en/docs/http/ngx_http_log_module.html#log_format)

I'd like to have a try if it's okay.

### PR DESCRIPTION
**Pull request by doujiang24**
Commit Message: access log: added new START_TIME_LOCAL & EMIT_TIME_LOCAL command
Additional Description:

Same as START_TIME & EMIT_TIME, but use local time zone.

close #35116 

Risk Level: low.
Testing: unit.
Docs Changes: added.
Release Notes: TODO.
Platform Specific Features: n/a.

### CODE REVIEW

**Comment by doujiang24** [comment]
PTAL cc @wbpcode Thanks.

**Comment by wbpcode** [comment]
In order to review this PR, we check the code of `DataFormatter` and find a bug. See https://github.com/envoyproxy/envoy/issues/35404

I will inclined to fix that bug first. And I think it will simply this PR.

/wait-any

File: source/common/common/utility.cc [hunk 1]
```diff
@@ -100,7 +102,7 @@ std::string DateFormatter::fromTime(const SystemTime& time) const {
   const std::chrono::seconds epoch_time_seconds =
       std::chrono::duration_cast<std::chrono::seconds>(epoch_time_ns);
 
-  const auto& item = cached_time.formatted.find(raw_format_string_);
+  const auto& item = cached_time.formatted.find(std::make_pair(raw_format_string_, local_time_));
```
Comment (wbpcode) [inline source/common/common/utility.cc:105]: This is a performance sensitive feature. It is hard to say that make a pair is free. If the format string length is larger then size of SSO, it will involve an additional heap allocation.

So I think it would be better to use two different maps to cache local time and utc time.

**Review by wbpcode — COMMENTED** [review]
Thanks for the contribution. And please also provide the comparation of the speed test results between the old implementation and the new one.

**Comment by doujiang24** [comment]
@wbpcode Thanks and sorry for the late.

Here are the benchmark results:
```
---------------------------------------------------------------------------------------------
Benchmark                                                   Time             CPU   Iterations
---------------------------------------------------------------------------------------------
# old
bmAccessLogDateTimeFormatter_stddev             252 ns          227 ns       100000
bmDateTimeFormatterWithSubseconds_stddev        193 ns          114 ns       100000

# new
bmAccessLogDateTimeFormatter_stddev             261 ns          233 ns       100000
bmDateTimeFormatterWithSubseconds_stddev        182 ns          114 ns       100000
```

File: source/extensions/filters/network/mongo_proxy/proxy.cc [hunk 2]
```diff
@@ -49,9 +49,9 @@ void AccessLog::logMessage(const Message& message, bool full,
       "{{\"time\": \"{}\", \"message\": {}, \"upstream_host\": \"{}\"}}\n";
 
   SystemTime now = time_source_.systemTime();
-  std::string log_line =
-      fmt::format(log_format, AccessLogDateTimeFormatter::fromTime(now), message.toString(full),
-                  upstream_host ? upstream_host->address()->asString() : "-");
+  std::string log_line = fmt::format(log_format, AccessLogDateTimeFormatter::fromTime(now, false),
+                                     message.toString(full),
+                                     upstream_host ? upstream_host->address()->asString() : "-");
```
Comment (wbpcode) [inline source/extensions/filters/network/mongo_proxy/proxy.cc:54]: Unnecessary change.

File: source/common/formatter/stream_info_formatter.h [hunk 3]
```diff
@@ -178,7 +178,9 @@ class SystemTimeFormatter : public StreamInfoFormatterProvider {
       std::function<absl::optional<SystemTime>(const StreamInfo::StreamInfo& stream_info)>;
   using TimeFieldExtractorPtr = std::unique_ptr<TimeFieldExtractor>;
 
-  SystemTimeFormatter(const std::string& format, TimeFieldExtractorPtr f);
+  SystemTimeFormatter(const std::string& format, TimeFieldExtractorPtr f, bool local_time);
+  SystemTimeFormatter(const std::string& format, TimeFieldExtractorPtr f)
+      : SystemTimeFormatter(format, std::move(f), false){};
```
Comment (wbpcode) [inline source/common/formatter/stream_info_formatter.h:183]: add a default value for the local_time to avoid additional constructor.

File: test/common/common/utility_test.cc [hunk 4]
```diff
@@ -594,13 +594,13 @@ TEST(StringUtil, removeCharacters) {
...
+  EXPECT_EQ("2018-04-03T23:06:09.000Z", AccessLogDateTimeFormatter::fromTime(time1, false));
   SystemTime time2(std::chrono::milliseconds(1522796769123));
-  EXPECT_EQ("2018-04-03T23:06:09.123Z", AccessLogDateTimeFormatter::fromTime(time2));
+  EXPECT_EQ("2018-04-03T23:06:09.123Z", AccessLogDateTimeFormatter::fromTime(time2, false));
   SystemTime time3(std::chrono::milliseconds(1522796769999));
-  EXPECT_EQ("2018-04-03T23:06:09.999Z", AccessLogDateTimeFormatter::fromTime(time3));
+  EXPECT_EQ("2018-04-03T23:06:09.999Z", AccessLogDateTimeFormatter::fromTime(time3, false));
   SystemTime time4(std::chrono::milliseconds(1522796768999));
-  EXPECT_EQ("2018-04-03T23:06:08.999Z", AccessLogDateTimeFormatter::fromTime(time4));
+  EXPECT_EQ("2018-04-03T23:06:08.999Z", AccessLogDateTimeFormatter::fromTime(time4, false));
```
Comment (wbpcode) [inline test/common/common/utility_test.cc:603]: unnecessary change. The fromTime will give a default false.

Comment (wbpcode) [reply test/common/common/utility_test.cc:603]: And please add some new tests for local time zone for both `DateFormatter` and `AccessLogDateTimeFormatter::fromTime`

**Review by wbpcode — COMMENTED** [review]
Thanks for the update. It's near there. Some comments are added. And please add a simple speed test in the utility_speed_test.cc for the AccessLogDateTimeFormatter::fromTime and new DateFormatter with local time zone.
(I guess it won't bring much effect because there is only some additional bool check, but this is a performance sensitive feature anyway)

**Review by wbpcode — APPROVED** [review]

**Comment by wbpcode** [comment]
LGTM. Thanks.
